File: learn/learn-algo/leetcode/1609/go/1609.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isEvenOddTree(root *TreeNode) bool {
	q := []*TreeNode{root}
	var prev int

	for level := 0; len(q) > 0; level++ {
		if isOdd(level) {
			prev = math.MaxInt32
		} else {
			prev = math.MinInt32
		}

		n := len(q)

		for _, v := range q {
			if isOdd(level) {
				if isOdd(v.Val) || v.Val >= prev {
					return false
				}
			} else {
				if isEven(v.Val) || v.Val <= prev {
					return false
				}
			}

			if v.Left != nil {
				q = append(q, v.Left)
			}

			if v.Right != nil {
				q = append(q, v.Right)
			}

			prev = v.Val
		}

		q = q[n:]

	}
	return true
}

func isOdd(n int) bool {
	return 1 == (n & 1)
}

func isEven(n int) bool {
	return 1 != (n & 1)
}

File: learn/learn-algo/leetcode/1609/c++/1609.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>
#include <math.h>
#include <set>
#include <map>
#include <unordered_map>
#include <unordered_set>
#include <queue>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


// INT_MAX 2147483647 10位数

typedef long long LL;

class Solution {
public:
    bool isOdd(LL n) {
        return n & 1;
    }

    bool isEven(LL n) {
        return !(n & 1);
    }

    bool isEvenOddTree(TreeNode *root) {
        queue<TreeNode *> q;

        q.push(root);
        int level = 0;

        while (!q.empty()) {
            int n = q.size();
            int preV = isOdd(level) ? INT_MAX : INT_MIN;
            while (n) {
                TreeNode *t = q.front();
                q.pop();

                if (isOdd(level)) {
                    if (!isEven(t->val) || t->val >= preV) {
                        cout << "l, v: " << level << " " << t->val << " odd" << endl;
                        return false;
                    }
                    preV = t->val;
                } else {
                    if (!isOdd(t->val) || t->val <= preV) {
                        cout << "l, v: " << level << " "<< t->val << " even" << endl;
                        return false;
                    }
                    preV = t->val;
                }

                if (t->left) q.push(t->left);
                if (t->right) q.push(t->right);
                n--;
            }
            level++;
        }
        return true;
    }
};


int main() {
    cout << endl;
    return 0;
}